ocaml/xapi/xapi_guest_agent.ml

```diff
@@ -88,22 +88,51 @@ let networks path (list: string -> string list) =
     let extract_network_keys eth =
       let iface_prefixes =
         ["eth"; "xenbr"; "eno"; "ens"; "emp"; "enx"] in
-      let string_after_prefix ~prefix str =
-        let prefix_len = String.length prefix in
-        let size = String.length str - prefix_len in
-        let after_prefix = String.sub str prefix_len size in
-        after_prefix
+      (* get vif id of the eth which own these mac addresses *)
+      let rec find_vif_id mac_list vifs =
+        match vifs with
+        | [] -> None
+        | vif_id :: rest ->
+          let mac_of_vif = (fun (xskey) -> match lookup xskey with
+            | Some xsval -> Some xsval
+            | None -> None)
+            (Printf.sprintf "device/vif/%s/mac" vif_id)
+          in
+          match mac_of_vif with
+          | None -> find_vif_id mac_list rest
+          | Some mac_of_vif -> 
+            if List.mem mac_of_vif mac_list then begin
+              Some vif_id
+            end else
+              find_vif_id mac_list rest
+      in
+      (* get all mac addresses of eth *)
+      let get_mac_list eth =
+        let key_prefix = Printf.sprintf "attr/%s/mac" eth in
+        let rec deopt acc = function
```

Comment:
I believe `deport x` returns the list of `Some` values. I think this would be clearer:

```
let depot options = options |> List.filter (fun x -> x <> None) |> List.map Opt.unbox
```

This could be also used directly at the call site below.